### PR TITLE
 DBZ-1128 Data could be lost from incompletely streamed tx after restart

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-1128
includes https://github.com/debezium/debezium/pull/780

### CODE REVIEW

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java [hunk 1]
```diff
@@ -93,6 +93,10 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
             LOGGER.info("Last position recorded in offsets is {}", lastProcessedPositionOnStart);
 
             TxLogPosition lastProcessedPosition = lastProcessedPositionOnStart;
+
+            // LSN should be increased for the first run only immediately after snapshot completion
+            // otherwise we might skip an incomplete transaction after restart
+            boolean shouldIncreaseFromLsn = offsetContext.isSnapshotCompleted();
```
Comment (grzegorz8) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java]: Just to be sure - what is the value of snapshotCompleted when the offset saved is like `{"commit_lsn":"000443d8:0073d0da:0005","change_lsn":"000443d8:0073d037:00a8"}`?

Comment (jpechane) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java]: @jpechane Yes, it is false. I've added two more test case to cover potential paths.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java [hunk 2]
```diff
@@ -103,15 +107,17 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
                     continue;
                 }
                 // There is no change in the database
-                if (currentMaxLsn.equals(lastProcessedPosition.getCommitLsn())) {
+                if (currentMaxLsn.equals(lastProcessedPosition.getCommitLsn()) && shouldIncreaseFromLsn) {
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java]: Why is `shouldIncreaseFromLsn` checked here, too?

Comment (jpechane) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java]: If a transaction was not processed completely and connector was restarted then the unfinished transaction would not be completed until a new transaction was written (increasing maxlsn).

**Comment by gunnarmorling** [comment]
LGTM, going to merge it.

@grzegorz8, this will be part of tomorrow's 0.9.1 release. If you had the chance to test it and report back whether it solves the issue on your end, that'd be great. Thanks a lot!

**Pull request merged**